File: fileops.py
"""
File manager + backup helpers. Every function here takes a `root` (a server's
data directory, controlled by us, never by the caller) and a `rel_path`
(user-supplied). safe_join() is the single chokepoint that prevents path
traversal -- nothing in this file should touch the filesystem with a path
that didn't come out of it.
"""
import datetime
import os
import shutil
import tarfile
from pathlib import Path

from werkzeug.utils import secure_filename


class PathSecurityError(Exception):
    pass


def safe_join(root: Path, rel_path: str) -> Path:
    root = root.resolve()
    rel_path = (rel_path or "").lstrip("/")
    candidate = (root / rel_path).resolve()
    try:
        candidate.relative_to(root)
    except ValueError:
        raise PathSecurityError(f"Path escapes server data directory: {rel_path!r}")
    return candidate


def list_dir(root: Path, rel_path: str):
    target = safe_join(root, rel_path)
    if not target.exists() or not target.is_dir():
        raise FileNotFoundError(rel_path)
    entries = []
    for entry in sorted(target.iterdir(), key=lambda p: (p.is_file(), p.name.lower())):
        try:
            stat = entry.stat()
            entries.append(
                {
                    "name": entry.name,
                    "is_dir": entry.is_dir(),
                    "size": stat.st_size if entry.is_file() else None,
                    "mtime": datetime.datetime.fromtimestamp(stat.st_mtime),
                }
            )
        except OSError:
            continue
    return entries


def read_text_file(root: Path, rel_path: str, max_bytes: int):
    target = safe_join(root, rel_path)
    if not target.is_file():
        raise FileNotFoundError(rel_path)
    if target.stat().st_size > max_bytes:
        raise ValueError("File too large to edit in the browser")
    return target.read_text(encoding="utf-8", errors="replace")


def write_text_file(root: Path, rel_path: str, content: str):
    target = safe_join(root, rel_path)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(content, encoding="utf-8")


def save_upload(root: Path, rel_dir: str, file_storage):
    filename = secure_filename(file_storage.filename)
    if not filename:
        raise ValueError("Invalid filename")
    target_dir = safe_join(root, rel_dir)
    target_dir.mkdir(parents=True, exist_ok=True)
    target = safe_join(root, os.path.join(rel_dir, filename))
    file_storage.save(target)
    return target


def delete_path(root: Path, rel_path: str):
    target = safe_join(root, rel_path)
    if target == root.resolve():
        raise PathSecurityError("Refusing to delete the server's root data directory")
    if target.is_dir():
        shutil.rmtree(target)
    elif target.exists():
        target.unlink()


def make_dir(root: Path, rel_path: str):
    target = safe_join(root, rel_path)
    target.mkdir(parents=True, exist_ok=True)


def resolve_download(root: Path, rel_path: str) -> Path:
    target = safe_join(root, rel_path)
    if not target.is_file():
        raise FileNotFoundError(rel_path)
    return target


# --- Backups -----------------------------------------------------------

def create_backup(backups_root: Path, server_slug: str, data_dir: Path) -> Path:
    dest_dir = backups_root / server_slug
    dest_dir.mkdir(parents=True, exist_ok=True)
    timestamp = datetime.datetime.utcnow().strftime("%Y%m%d-%H%M%S")
    archive_path = dest_dir / f"{server_slug}-{timestamp}.tar.gz"
    with tarfile.open(archive_path, "w:gz") as tar:
        tar.add(data_dir, arcname=".")
    return archive_path


def list_backups(backups_root: Path, server_slug: str):
    dest_dir = backups_root / server_slug
    if not dest_dir.exists():
        return []
    backups = []
    for entry in sorted(dest_dir.glob("*.tar.gz"), reverse=True):
        stat = entry.stat()
        backups.append(
            {
                "filename": entry.name,
                "size_mb": round(stat.st_size / (1024 * 1024), 2),
                "mtime": datetime.datetime.fromtimestamp(stat.st_mtime),
            }
        )
    return backups


def resolve_backup(backups_root: Path, server_slug: str, filename: str) -> Path:
    filename = secure_filename(filename)
    target = safe_join(backups_root / server_slug, filename)
    if not target.is_file():
        raise FileNotFoundError(filename)
    return target


def delete_backup(backups_root: Path, server_slug: str, filename: str):
    target = resolve_backup(backups_root, server_slug, filename)
    target.unlink()

File: internal/fileops/fileops.go
// Package fileops is the file manager + backup helper layer, replacing
// fileops.py. safe_join (Join here) is the single chokepoint that prevents
// path traversal -- nothing in this package should touch the filesystem with
// a path that didn't come out of it.
package fileops

import (
	"archive/tar"
	"compress/gzip"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

var ErrPathSecurity = errors.New("path escapes server data directory")

type Entry struct {
	Name  string
	IsDir bool
	Size  int64 // -1 for directories
	MTime time.Time
}

type Backup struct {
	Filename string
	SizeMB   float64
	MTime    time.Time
}

// Join resolves rel under root and guarantees the result stays within root.
func Join(root, rel string) (string, error) {
	absRoot, err := filepath.Abs(root)
	if err != nil {
		return "", err
	}
	absRoot, err = filepath.EvalSymlinks(absRoot)
	if err != nil {
		// root may not exist yet (e.g. mkdir on a fresh subtree); fall back to the
		// non-symlink-resolved absolute path rather than failing outright.
		absRoot, _ = filepath.Abs(root)
	}
	rel = strings.TrimLeft(rel, "/")
	candidate := filepath.Join(absRoot, rel)
	candidate = filepath.Clean(candidate)

	rootWithSep := absRoot
	if !strings.HasSuffix(rootWithSep, string(filepath.Separator)) {
		rootWithSep += string(filepath.Separator)
	}
	if candidate != absRoot && !strings.HasPrefix(candidate, rootWithSep) {
		return "", ErrPathSecurity
	}
	return candidate, nil
}

func ListDir(root, rel string) ([]Entry, error) {
	target, err := Join(root, rel)
	if err != nil {
		return nil, err
	}
	info, err := os.Stat(target)
	if err != nil || !info.IsDir() {
		return nil, os.ErrNotExist
	}
	dirEntries, err := os.ReadDir(target)
	if err != nil {
		return nil, err
	}
	entries := make([]Entry, 0, len(dirEntries))
	for _, de := range dirEntries {
		fi, err := de.Info()
		if err != nil {
			continue
		}
		size := int64(-1)
		if !de.IsDir() {
			size = fi.Size()
		}
		entries = append(entries, Entry{
			Name:  de.Name(),
			IsDir: de.IsDir(),
			Size:  size,
			MTime: fi.ModTime(),
		})
	}
	// Directories first, then files, both alpha (matches Python: key=(is_file, name.lower())).
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].IsDir != entries[j].IsDir {
			return entries[i].IsDir // dirs before files
		}
		return strings.ToLower(entries[i].Name) < strings.ToLower(entries[j].Name)
	})
	return entries, nil
}

func ReadTextFile(root, rel string, maxBytes int64) (string, error) {
	target, err := Join(root, rel)
	if err != nil {
		return "", err
	}
	info, err := os.Stat(target)
	if err != nil || info.IsDir() {
		return "", os.ErrNotExist
	}
	if info.Size() > maxBytes {
		return "", fmt.Errorf("file too large to edit in the browser")
	}
	data, err := os.ReadFile(target)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func WriteTextFile(root, rel, content string) error {
	target, err := Join(root, rel)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
		return err
	}
	return os.WriteFile(target, []byte(content), 0o644)
}

var secureFilenameStrip = regexp.MustCompile(`[^A-Za-z0-9_.-]+`)

// SecureFilename mirrors werkzeug.utils.secure_filename closely enough for
// our purposes: strip directory components, collapse anything unsafe.
func SecureFilename(name string) string {
	name = filepath.Base(name)
	name = secureFilenameStrip.ReplaceAllString(name, "_")
	name = strings.Trim(name, "._")
	return name
}

func SaveUpload(root, relDir string, fileHeader *multipart.FileHeader) (string, error) {
	filename := SecureFilename(fileHeader.Filename)
	if filename == "" {
		return "", fmt.Errorf("invalid filename")
	}
	targetDir, err := Join(root, relDir)
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(targetDir, 0o755); err != nil {
		return "", err
	}
	target, err := Join(root, filepath.Join(relDir, filename))
	if err != nil {
		return "", err
	}
	src, err := fileHeader.Open()
	if err != nil {
		return "", err
	}
	defer src.Close()
	dst, err := os.Create(target)
	if err != nil {
		return "", err
	}
	defer dst.Close()
	if _, err := io.Copy(dst, src); err != nil {
		return "", err
	}
	return target, nil
}

func DeletePath(root, rel string) error {
	target, err := Join(root, rel)
	if err != nil {
		return err
	}
	absRoot, _ := filepath.Abs(root)
	if target == absRoot {
		return fmt.Errorf("%w: refusing to delete the server's root data directory", ErrPathSecurity)
	}
	info, err := os.Stat(target)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	if info.IsDir() {
		return os.RemoveAll(target)
	}
	return os.Remove(target)
}

func MakeDir(root, rel string) error {
	target, err := Join(root, rel)
	if err != nil {
		return err
	}
	return os.MkdirAll(target, 0o755)
}

func ResolveDownload(root, rel string) (string, error) {
	target, err := Join(root, rel)
	if err != nil {
		return "", err
	}
	info, err := os.Stat(target)
	if err != nil || info.IsDir() {
		return "", os.ErrNotExist
	}
	return target, nil
}

// --- Backups -----------------------------------------------------------

func CreateBackup(backupsRoot, serverSlug, dataDir string) (string, error) {
	destDir := filepath.Join(backupsRoot, serverSlug)
	if err := os.MkdirAll(destDir, 0o755); err != nil {
		return "", err
	}
	timestamp := time.Now().UTC().Format("20060102-150405")
	archivePath := filepath.Join(destDir, fmt.Sprintf("%s-%s.tar.gz", serverSlug, timestamp))

	f, err := os.Create(archivePath)
	if err != nil {
		return "", err
	}
	defer f.Close()
	gz := gzip.NewWriter(f)
	defer gz.Close()
	tw := tar.NewWriter(gz)
	defer tw.Close()

	err = filepath.Walk(dataDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(dataDir, path)
		if err != nil {
			return err
		}
		if rel == "." {
			return nil
		}
		// arcname="." in the Python version means entries are stored relative to
		// the data dir root (no leading server-slug/ component) -- match that.
		hdr, err := tar.FileInfoHeader(info, "")
		if err != nil {
			return err
		}
		hdr.Name = filepath.ToSlash(rel)
		if info.IsDir() {
			hdr.Name += "/"
		}
		if err := tw.WriteHeader(hdr); err != nil {
			return err
		}
		if info.Mode().IsRegular() {
			src, err := os.Open(path)
			if err != nil {
				return err
			}
			defer src.Close()
			if _, err := io.Copy(tw, src); err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		return "", err
	}
	return archivePath, nil
}

func ListBackups(backupsRoot, serverSlug string) ([]Backup, error) {
	destDir := filepath.Join(backupsRoot, serverSlug)
	entries, err := os.ReadDir(destDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var backups []Backup
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".tar.gz") {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		backups = append(backups, Backup{
			Filename: e.Name(),
			SizeMB:   float64(info.Size()) / (1024 * 1024),
			MTime:    info.ModTime(),
		})
	}
	sort.Slice(backups, func(i, j int) bool { return backups[i].Filename > backups[j].Filename })
	return backups, nil
}

func ResolveBackup(backupsRoot, serverSlug, filename string) (string, error) {
	filename = SecureFilename(filename)
	target, err := Join(filepath.Join(backupsRoot, serverSlug), filename)
	if err != nil {
		return "", err
	}
	info, err := os.Stat(target)
	if err != nil || info.IsDir() {
		return "", os.ErrNotExist
	}
	return target, nil
}

func DeleteBackup(backupsRoot, serverSlug, filename string) error {
	target, err := ResolveBackup(backupsRoot, serverSlug, filename)
	if err != nil {
		return err
	}
	return os.Remove(target)
}
